### ISSUE TITLE
Why the return of predict and predict generator are different?

### ISSUE THREAD

**Issue by guirighetto** [keras-team/keras#3477]
Hello, I'm using lately ImageDataGenerator to be able to use dataser larger. So I am trying to replicate the same example using dataArgumentation, I get the same accuracy for the two examples, but the return of model.predict and model.predict_generator are completely different. Does anyone know why this happens? Remembering that I know they can be different but in this case is completely different. Below the two examples with the same architecture.

dataArgumentation:

```
`   model.compile(loss="categorical_crossentropy", optimizer="Adadelta",metrics=['accuracy'])
    train_datagen = ImageDataGenerator(rescale=1./255)

    test_datagen = ImageDataGenerator(rescale=1./255)

    train_generator = train_datagen.flow_from_directory(
        trainPath,
        target_size=(blockWidth, blockHeight),
        color_mode='grayscale',
        batch_size=batchSize,
        class_mode='categorical',
        shuffle=True)

    validation_generator = test_datagen.flow_from_directory(
        testPath,
        target_size=(blockWidth, blockHeight),
        color_mode='grayscale',
        batch_size=batchSize,
        class_mode='categorical',
        shuffle=False)

    model.fit_generator(
        train_generator,
        samples_per_epoch=10500,
        nb_epoch=epoch,
        validation_data=validation_generator,
        nb_val_samples=5250)

    scoreSeg = model.evaluate_generator(validation_generator,5250)
    print("Accuracy = ",scoreSeg[1])
    predict = model.predict_generator(validation_generator,5250)

```

No dataArgumetation:

```
def loadDatabase():

    numSamplesTrain = float(numImgClass*(float(train)/100))
    numSamplesTrain = round(numSamplesTrain)

    dataTrain = []
    labelTrain = []
    dataTest = []
    labelTest = []
    filesTest = []
    filesCount = 0
    patchesCount = 0

    for c in range(1,classes+1):
        filesTest.append([])

        for s in range(1,numImgClass+1):

            if(s < numSamplesTrain+1):
                folderTrainTest = 'Treino/'
            else:
                folderTrainTest = 'Teste/'

            for b in range(1,numblock+1):
                nameImg = preName + str(c).zfill(5) + sep + str(s) + sep + str(b)
                folderClass = preName + str(c).zfill(5) + '/'
                fullPathImg = pathImages + folderClass + folderTrainTest + nameImg + imagesFormat
                image = plt.imread(fullPathImg)

                image = image[np.newaxis]

                if(folderTrainTest == 'Treino/'):
                    dataTrain.append(image)
                    labelTrain.append(c-1)
                else:
                    dataTest.append(image)
                    labelTest.append(c-1)
                    filesTest[filesCount].append(patchesCount)
                    patchesCount += 1
        filesCount+=1

    #train
    dataTrain = np.array(dataTrain)
    labelTrain = np.array(labelTrain)

    LUT = np.arange(len(dataTrain), dtype=int)
    random.shuffle(LUT)
    randomDataTrain = dataTrain[LUT]
    randomLabelTrain = labelTrain[LUT]

    X_Train = randomDataTrain.astype("float32")/255
    Y_Train = np_utils.to_categorical(randomLabelTrain, classes)

    #test       
    dataTest = np.array(dataTest)
    labelTest = np.array(labelTest)
    filesTest = np.array(filesTest)

    X_Test = dataTest.astype("float32")/255
    Y_Test = np_utils.to_categorical(labelTest, classes)

    print("Number samples train: ",numSamplesTrain*numblock*classes)
    print("Number samples test: ",(numImgClass-numSamplesTrain)*numblock*classes)       

    return X_Train, Y_Train, X_Test, Y_Test, filesTest

main():
    X_Train, Y_Train, X_Test, Y_Test, filesTest = loadDatabase()

    model = buildModel()

    model.compile(loss="categorical_crossentropy", optimizer="Adadelta",metrics=['accuracy'])
    model.fit(X_Train, Y_Train, nb_epoch=epoch, batch_size=batchSize, verbose=1, validation_data=(X_Test, Y_Test))
    scoreSeg = model.evaluate(X_Test, Y_Test, verbose=1)
    predict = model.predict(X_Test, verbose=1)
```

**Comment by myway0101** [comment]
Hi 
Did you get any solution for this??

**Comment by daavoo** [comment]
I'm not sure but a after a quick reading of your code I think that the problem is that you are "shuffling" the test set in your example of "No dataArgumetation:"

```python
LUT = np.arange(len(dataTrain), dtype=int)
random.shuffle(LUT)
randomDataTrain = dataTrain[LUT]
randomLabelTrain = labelTrain[LUT]
```
And this is making your model to predict according to this random shuffle. So your first prediction do not correspond to your first file.

While on the other hand:

```
validation_generator = test_datagen.flow_from_directory(
        testPath,
        target_size=(blockWidth, blockHeight),
        color_mode='grayscale',
        batch_size=batchSize,
        class_mode='categorical',
        shuffle=False)
```
You are setting **shuffle=False** so in this case the orden is correct.

**Comment by guirighetto** [comment]
No, to solve this problem I implemented algorithm to load batch images into memory, therefore I dont used dataArgumentation.

**Comment by sysid** [comment]
I have the same problem: predict and predict generator resulting in very different results. Using keras.applications.vgg16. 'predict' results are correct whereas 'predict_generator' results are wrong. 

Can't find the problem.

**Comment by Aakash282** [comment]
@sysid Could you explain why we shouldn't use rescale? How else will the model know to do that rescale on the test data. Furthermore, this doesn't really explain what is the difference between these two functions. Please elaborate if you remember

**Comment by sakvaua** [comment]
@Aakash282 VGG was trained on demeaned but not rescaled data.

**Comment by Aakash282** [comment]
ok sure. But i think this problem extends past that one model/dataset. I trained a cnn using a generator and validation_generator. So far, the only way to get remotely consistent results is to use a generator for my test data but I want to actually know which results map to which images. I tried using batch but that didn't really work either.

**Comment by ardianumam** [comment]
@guirighetto  : I also got same problem. After some trials, then I know that the batch size should be a number that can divide the size of training data and validation data. For example, training data : 1000 and validation data: 100, batch size that is allowed for example is 10. If you use bath size for example 15, your training data will be only 990 and validation data will be only 90, which are biggest number that are divisible by 15. I think this issue can be resolved for next Keras version :) Cc: @fchollet

**Comment by sakvaua** [comment]
@ardianumam you can also increment number of steps by one and select first n elements, where n is number of training samples

**Comment by ardianumam** [comment]
@sakvaua : Thanks, Man! It works. I have one more question. For the case I have 100 validation images, and batch_size : 15, then I do "bottleneck_features_validation = model.predict_generator(generator,
 ((validation_samples // batch_size)+1))", where "validation_samples // batch_size = 100//15 = 6". If without "+1" I only get 6x15 = 90. If with "+1", then I get 7x15= 105, and since my val images only 100, thus, is it automatically stop at 100? I mean, not become 105 by taking any 5 images random again.

**Comment by yinleon** [comment]
I read over these comments, and I am getting the same issue for vgg16.

The code block uses pre-trained vgg16 to return features for `train_img` (N=1798).
The output of <i>predict_generator</i> (`pred_0`) is different than <i>predict</i> (`pred_0`)!

```
batch_size = 64

model = VGG16(input_shape=train_img[0].shape, 
              weights='imagenet', 
              include_top=False)

for layer in model.layers:
    layer.trainable = False

gen = ImageDataGenerator()

pred_0 = model.predict_generator(
    gen.flow(train_img, batch_size=batch_size),
    steps=(len(train_img) // batch_size) + 1, # 25
    verbose=1
)
pred_1 = model.predict(train_img, verbose=1)

np.array_equal(pred_0[0], pred_1[0]) # False
pred_0.shape # (1598, 7, 7, 512)
pred_1.shape # (1598, 7, 7, 512)
```

I'm using Python 3.6 and Keras 2.06

**Comment by EnricoBeltramo** [comment]
I had similar issue and after a few of work I found that to have a correct match I had to:
- remove all transformations (i.e. rescale)
- set shuffle = False, in order to have same sequences then original (otherwise isn't clear the order of samples vs result)
- batch_size same of number of samples, in order to have exact number of samples as original 
After those change, the comparison with "manual" provided samples and prediction match exactly.

**Comment by YasserMustafa** [comment]
Hi All,

I had the same problem but I solved it! The clue is to find an integer number for the batch_size that could divide both of nb_train_samples and nb_validation_samples; i.e.,
nb_train_samples//batch_size = 0  and nb_validation_samples//batch_size = 0.
As an example, if:
nb_train_samples = 2400
nb_validation_samples = 1200
then:
batch_size = 4 or 6 or....... or 48 or ..... .

**Comment by soumendra** [comment]
The output from `predict()` and `predict_generator()` are actually identical, but they look different because they are labelled differently. You are providing the labels to `predict()`, and `predict_generator()` is inferring the labels (because it's using `flow_from_directory()` instead of `flow()`) from the directory structure of training data. These can be very different, [as I explain in this blogpost](https://medium.com/@soumendra/keras-a-thing-you-should-know-about-keras-if-you-plan-to-train-a-deep-learning-model-on-a-large-fdd63ce66bd2).

Here is some code to convert the mapping from `predict_generator()` to the one you supplied to `predict()` so that you can see that they are identical:

```python
import numpy as np
predictions = model.predict_generator(self.test_generator)
predictions = np.argmax(predictions, axis=-1) #multiple categories

label_map = (train_generator.class_indices)
label_map = dict((v,k) for k,v in label_map.items()) #flip k,v
predictions = [label_map[k] for k in predictions]
```

**Comment by ziadloo** [comment]
If you are getting unexpected results from `predict_generator()` look into the following cases:

1. Make sure that your generator is not augmenting validation/test data when instantiating `ImageDataGenerator`
1. Make sure that `shuffle` is false when calling `flow` or `flow_from_directory` for test data
1. Make sure that the files are read in the same order as you want by `flow_from_directory`

### Augmenting the data

When you are reading images using `ImageDataGenerator`, there are arguments to the class constructor that activate the "augmentation" later when the images are returned as sample points. Here are a bunch of them:

- `rotation_range`
- `width_shift_range`
- `height_shift_range`
- `shear_range`
- `zoom_range`
- `horizontal_flip`

While augmentation is something you want to look into for training data, you don't want it for validation and/or test data. So make sure you are not going to set these parameters for the instance of the `ImageDataGenerator` that you are going to use to read validation and/or test data.

### Shuffling data

Again, this parameter is something useful only for training data and not for validation and/or test data (as for validation it really does not matter, but it does no good either). What you need to know is that for training and validation, the sample points are accompanies with labels so when they are shuffled, their coresponding labels are shuffled with the sample points so the bond is not broken. But in case of test data, the ralation between sample points and their labels is based on their order (at least usually). So if you shuffle the test data, you are breaking that bond.

### How files are read

This I found through test and debugging and it's again related to the order of test data. Consider your data points are images and they are stored on hard disk as files. Usually, the order of files (based on their filenames) is assumed and the labels are stored (for instance all in a CSV file) following the exact same order as filenames. But there's a problem here. Ordering of filenames (strings in general) could be done in different ways. Consider the following two cases:

1. `1.jpg`
1. `2.jpg`
1. `10.jpg`

And

1. `1.jpg`
1. `10.jpg`
1. `2.jpg`

Your OS's file manager will most likely use the first ordering of files when you open the container folder. But `flow_from_directory` method will read them ordered with the second method. As I'm sure you can see it yourself, this will lead to the same problem as the one before. The bond between data points and their labels will be broken. The solution to this is to rename the files so `flow_from_directory` will read them as the first case:

1. `01.jpg`
1. `02.jpg`
1. `10.jpg`

You can do this using a simple peice of code. But you need to know how many files you have beforehand so you can pad `0`s to the left of all filename according to that.

After applying these steps, I managed to get the correct output out of `predict_generator`.

**Comment by atortoricimontaperto** [comment]
Still many moving parts on this. There may be many contributions to the error and it is hard to identify the one. It is annoying to have to set no rescale in the test data, because often the training is performed with rescaling and if we want to rescale also the test data outside the Keras sequential API e.g. using some Image libraries such as PIL or OpenCV, it is extra work. So I tried to fix everything to good, except the rescaling and see how it did go.
For a binary classification (e.g. cats/dogs) I have tested the generator over a train directory of 2000 images, a validation directory of 800 images, a test directory of 112 unlabeled images (that is, there is only 1 subdirectory). All these directories contain a number of images divisible exactly by 16.  
So here is the main configuration parameters, starting with the ImageDataGenerator:

# various transformations, including rescaling, for training
train_datagen = ImageDataGenerator(
        rescale=1./255,
        shear_range=0.2,
        zoom_range=0.2,
        horizontal_flip=True)

# only rescaling for validation
valid_datagen = ImageDataGenerator(rescale=1./255)

# only rescaling for test
test_datagen = ImageDataGenerator(rescale=1./255)

# files directories
traindir = 'C:/myDocs/Train/'
validdir = 'C:/MyDocs/Valid/'
testdir = 'C:/MyDocs/Test/'

# iterator, for validation data
train_iter = train_datagen.flow_from_directory(
        traindir,
        target_size=(150, 150),  # all images will be resized to 150x150
        batch_size=16, # 2000/16=125 exact!
        class_mode='binary') 

# this is a similar iterator, for validation data
validation_iter = valid_datagen.flow_from_directory(
        validdir,
        target_size=(150, 150),
        batch_size=16, # 800/16=50 exact!
        class_mode='binary')

test_iter = test_datagen.flow_from_directory(
        testdir,
        target_size=(150, 150),
        shuffle=False, # we need to avoid shuffle if we want to preserve the order of test
        batch_size=16, # 112/16=7 exact!
        class_mode=None)  # this is None for an unlabeled directory

# Learning the model
model.fit_generator(
        train_iter,
        steps_per_epoch=2000 / 16    #  exact
        epochs=50,
        validation_data=validation_iter,
        validation_steps=800 / 16)     # exact

We learn say to 70% accuracy or more on the validation data.
Now we try the predict generator on the test dataset of 112 images: 112/16 = 7

test_iter.reset() # Just in case
mypred_using_generator=model.predict_generator(test_iter,verbose=0, steps=7)     # 7x16=112 exact

I have used an order friendly file name:
001.jpg, 002.jpg,...,112.jpg  so the predictions should refer to this order. If I print test_iter.filenames I get precisely the order 001.jpg, 002.jpg,...,112.jpg
But,... when I verify manually 

model.predict(image001),
model.predict(image002)
…
model.predict(image112)

these predictions are different from predict_generator predictions on the images 001.jpg, 001.jpg, etc.. So let's summarize the good practices followed:
- shuffle=False
- batch_size: always exact, the batches always contain 16 images
- file naming of images: in the same order of test_iter.filenames
- iterator reset: OK
- Rescaling: **KO**, not switched off.

If I try now changing test_datagen = ImageDataGenerator(rescale=1./255)
with
test_datagen = ImageDataGenerator()

then the match between predict_generator and the manual predict is perfect.
It is interesting that the match is perfect, even if the train and the validation datagen where configured with rescaling and the model was learned on rescaled data.

So, it seems that we must avoid rescaling on test unlabeled data, but the no-rescaling, after all, is not modifying substantially the predictions performances.
Initially, I thought that using ImageDataGeneration with anything (as an example with rescaling), was shuffling the data (irrespectively that we specify False), or that a hidden ordering method of the files in the directories was mixing the results of the predict_generator. But that cannot be the case, as with predict-generator (and rescaling) the number of predictions for a given class is different of the number of prediction for the same class in the manual case, so it is not really a problem of not having the correct order on the predictions out of predict_generator (in other words it is not an issue of mapping incorrectly the predictions with the file name), the predictions are different overall!

Conclusion: Follow all the advices cleverly identified for this issue by many people, including (sigh) to switch off rescaling.

**Comment by cgebbe** [comment]
A small addition:

1. As @atortoricimontaperto says (thanks!), rescaling can be used with predict_generator, when you also take into account the rescaling in the manual predict

2. In addition: The size of the validation data set does not HAVE to be an exact multiple of the batch size (although it makes it easier). In that case, you can set "shuffle=False" and call  "mygen.reset()" which sets "mygen.batch_index=0" before calling "predict_generator(mygen). Then, the generator should start from index 0 again and iterate alphabetically. Worked for me at least. Before, the results gave me lots of headaches, since the generator started at some other index so that the predicted labels did not match the true labels at all...

**Comment by shaoeChen** [comment]
@cgebbe hi dear.
i work for me.
you solve my confuse.
thank you.

**Comment by messerzen** [comment]
Thanks @ziadloo and @cgebbe, the tips worked for me.
I created a simple code to rename files, since my dataset is too large. 

The link is below if the function interests someone:

https://github.com/messerzen/Capstone/blob/master/rename_function.py

**Comment by magicknight** [comment]
Dont fotget to set shuffle=False, Bros!

**Comment by buikhizarkhan** [comment]
> I have the same problem: predict and predict generator resulting in very different results. Using keras.applications.vgg16. 'predict' results are correct whereas 'predict_generator' results are totally different and wrong. I debugged input from ImageDataGenerator and it looks ok.
> 
> Problem found: Must not use rescale=1./255 in ImageDataGenerator.

I would be grateful if you please explain/elaborate, why must not use rescale-1./255 in ImageDataGenerator??
